File: internal/store/store.go
package store

import (
	"fmt"
	"os"
	"strings"

	"github.com/sakshamgupta05/cmdvault/internal/config"
	"gopkg.in/yaml.v3"
)

// Command represents a stored shell command
type Command struct {
	Collection  string
	Name        string   `yaml:"name"`
	Description string   `yaml:"description"`
	Command     string   `yaml:"command"`
	Tags        []string `yaml:"tags"`
}

type Collection struct {
	Commands []Command `yaml:"commands"`
}

func ListCollections() ([]string, error) {
	// List all collections by finding all YAML files in the commands directory
	collections := make([]string, 0)
	dirs := config.GetCollectionDirs()
	for _, dir := range dirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading commands directory: %v\n", err)
			return nil, err
		}

		for _, entry := range entries {
			if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".yaml") {
				collections = append(collections, strings.TrimSuffix(entry.Name(), ".yaml"))
			}
		}
	}

	return collections, nil
}

func GetAllCommands() ([]Command, error) {
	collections, err := ListCollections()
	if err != nil {
		return nil, err
	}

	var commands []Command
	for _, collection := range collections {
		cmds, err := GetCommands(collection)
		if err != nil {
			return nil, err
		}

		commands = append(commands, cmds...)
	}

	return commands, nil
}

// GetCommands returns all commands from the specified collection
func GetCommands(collection string) ([]Command, error) {
	collectionPath, err := config.GetCollectionPath(collection)
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(collectionPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: could not read file %s: %v\n", collectionPath, err)
		return nil, err
	}

	var collectionStruct Collection
	if err := yaml.Unmarshal(data, &collectionStruct); err != nil {
		fmt.Fprintf(os.Stderr, "Error: could not parse file %s: %v\n", collectionPath, err)
		return nil, err
	}
	for i := range collectionStruct.Commands {
		collectionStruct.Commands[i].Collection = collection
	}

	return collectionStruct.Commands, nil
}

// SearchCommands searches for commands matching the search term
func SearchCommands(searchTerm string) ([]Command, error) {
	commands, err := GetAllCommands()
	if err != nil {
		return nil, err
	}

	if searchTerm == "" {
		return commands, nil
	}

	searchTerm = strings.ToLower(searchTerm)
	var results []Command
	for _, cmd := range commands {
		if strings.Contains(strings.ToLower(cmd.Collection)+" "+strings.ToLower(cmd.Name), searchTerm) {
			results = append(results, cmd)
			continue
		}

		// Search tags
		for _, tag := range cmd.Tags {
			if strings.Contains(strings.ToLower(tag), searchTerm) {
				results = append(results, cmd)
				break
			}
		}
	}

	return results, nil
}

File: src/store.rs
use anyhow::{Context, Result};
use serde::Deserialize;
use std::fs;
use std::path::PathBuf;

use crate::config;

#[derive(Deserialize, Clone, Debug)]
pub struct Parameter {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub optional: bool,
    #[serde(default, rename = "defaultValue")]
    pub default_value: String,
}

#[derive(Deserialize, Clone, Debug)]
pub struct Command {
    #[serde(skip)]
    pub collection: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub command: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub parameters: Vec<Parameter>,
}

#[derive(Deserialize)]
pub struct Collection {
    #[serde(default)]
    pub commands: Vec<Command>,
}

/// List all collection names by scanning collection directories for .yaml files.
pub fn list_collections(collection_dirs: &[PathBuf]) -> Result<Vec<String>> {
    let mut collections = Vec::new();

    for dir in collection_dirs {
        let entries = fs::read_dir(dir)
            .with_context(|| format!("Error reading commands directory: {}", dir.display()))?;

        for entry in entries {
            let entry = entry?;
            let path = entry.path();
            if !path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if name.ends_with(".yaml") {
                        collections.push(name.trim_end_matches(".yaml").to_string());
                    }
                }
            }
        }
    }

    Ok(collections)
}

/// Get all commands from a specific collection.
pub fn get_commands(collection_dirs: &[PathBuf], collection: &str) -> Result<Vec<Command>> {
    let collection_path = config::get_collection_path(collection_dirs, collection)?;

    let data = fs::read_to_string(&collection_path)
        .with_context(|| format!("Error: could not read file {}", collection_path.display()))?;

    let mut col: Collection = serde_yaml::from_str(&data)
        .with_context(|| format!("Error: could not parse file {}", collection_path.display()))?;

    for cmd in &mut col.commands {
        cmd.collection = collection.to_string();
    }

    Ok(col.commands)
}

/// Get all commands from all collections.
pub fn get_all_commands(collection_dirs: &[PathBuf]) -> Result<Vec<Command>> {
    let collections = list_collections(collection_dirs)?;
    let mut all_commands = Vec::new();

    for collection in &collections {
        let cmds = get_commands(collection_dirs, collection)?;
        all_commands.extend(cmds);
    }

    Ok(all_commands)
}

/// Search commands by term (case-insensitive substring match on collection+name and tags).
pub fn search_commands(collection_dirs: &[PathBuf], search_term: &str) -> Result<Vec<Command>> {
    let commands = get_all_commands(collection_dirs)?;

    if search_term.is_empty() {
        return Ok(commands);
    }

    let term = search_term.to_lowercase();
    let mut results = Vec::new();

    for cmd in commands {
        let name_match = format!("{} {}", cmd.collection, cmd.name)
            .to_lowercase()
            .contains(&term);

        if name_match {
            results.push(cmd);
            continue;
        }

        let tag_match = cmd
            .tags
            .iter()
            .any(|tag| tag.to_lowercase().contains(&term));

        if tag_match {
            results.push(cmd);
        }
    }

    Ok(results)
}
